scripts/cluster-install/openstack_connection.py

```diff
@@ -0,0 +1,127 @@
+import time
+import traceback
+import openstack
+
+
+def _get_client_config(config):
+    client_config = {
+        'auth_url': config['auth_url'],
+        'username': config['username'],
+        'password': config['password'],
+        'region_name': config['region_name'],
+        'project_name': config['tenant_name'],
+        'user_domain_name': 'Default',
+        'project_domain_id': 'Default'
+    }
+    return client_config
+
+
+def _get_resource_config(config, name):
+    resource_config = {
+        'name': name,
+        'image_id': '70de1e0f-2951-4eae-9a8f-05afd97cd036',  # centos image_id
+        'flavor_id': '3',  # m1.medium id
+        'networks': [{'uuid': config['network_id']}],
+        'key_name': config['key_name']
+    }
+    return resource_config
+
+
+def _create_sec_group_and_open_all_ports(connection):
+    sec_group = connection.network.create_security_group(
+        name='cluster-sec-group')
+    connection.network.create_security_group_rule(
+        security_group_id=sec_group.id,
+        direction='ingress',
+        remote_ip_prefix='0.0.0.0/0',
+        protocol='tcp',
+        port_range_max='65535',
+        port_range_min='1',
+        ethertype='IPv4')
+    return sec_group
+
+
+def _get_private_ip(connection, server_id):
+    tmp_server = connection.get_server(server_id, detailed=True)
+    start_time = time.time()
+    while tmp_server.private_v4 == '':
+        tmp_server = connection.get_server(server_id, detailed=True)
+        end_time = time.time()
+        if (end_time-start_time) > 20 and tmp_server.private_v4 == '':
+            raise Exception('Error in the Openstack SDK')  # Timeout
+    return tmp_server.private_v4
+
+
+def _create_floating_ip(connection, config, resource_config):
+    return connection.create_floating_ip(network=config['gateway_net_id'],
+                                         server=resource_config)
+
+
+def _add_floating_ip_to_server(connection, server_id, floating_ip_address):
+    connection.compute.add_floating_ip_to_server(server_id,
+                                                 floating_ip_address)
+
+
+def _create_vm(connection, sec_group, config, instance, servers_ids, logging):
```

Comment:
This isn't a private method really is it?

If you're looking for somewhere to make some parallelised speed improvements, anywhere using this and the delete_vm method are probably good places to look. Since they're functions already you should be able to easily pass them into a thread pool or similar.

Reply:
Great idea, I changed it.